sycl/include/sycl/ext/intel/experimental/pipes.hpp

```diff
@@ -11,29 +11,135 @@
 #include "fpga_utils.hpp"
 #include <CL/__spirv/spirv_ops.hpp>
 #include <CL/__spirv/spirv_types.hpp>
+#include <sycl/context.hpp>
+#include <sycl/device.hpp>
+#include <sycl/ext/intel/experimental/pipe_properties.hpp>
 #include <sycl/ext/oneapi/properties/properties.hpp>
+#include <sycl/queue.hpp>
 #include <sycl/stl.hpp>
 #include <type_traits>
 
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+#include <xpti/xpti_data_types.h>
+#include <xpti/xpti_trace_framework.hpp>
+#endif
+
 namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
-namespace ext::intel::experimental {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+namespace detail {
+template <typename Properties, typename PropertyKey, typename Cond = void>
+struct ValueOrDefault {
+  template <typename ValT> static constexpr ValT get(ValT Default) {
+    return Default;
+  }
+};
+
+template <typename Properties, typename PropertyKey>
+struct ValueOrDefault<
+    Properties, PropertyKey,
+    std::enable_if_t<
+        sycl::ext::oneapi::experimental::is_property_list_v<Properties> &&
+        Properties::template has_property<PropertyKey>()>> {
+  template <typename ValT> static constexpr ValT get(ValT) {
+    return Properties::template get_property<PropertyKey>().value;
+  }
+};
+} // namespace detail
```

Comment:
Thanks @steffenlarsen. I moved `ValueOrDefault` to sycl/include/sycl/ext/oneapi/properties/properties.hpp instead of property_utils.hpp. Because the `is_property_list_v` used in the `ValueOrDefault` is defined in the properties.hpp, if I move `ValueOrDefault` to `property_utils.hpp`, I would have to include the `properties.hpp` in the `property_utils.hpp`, however, because of `properties.hpp` already includes `property_utils.hpp`, it creates circular dependency between  `properties.hpp` and `properties_utils.hpp`.

Please let me know if this make senses and I missed anything.

Reply:
I see. Another option is to move into another level of utility headers, but since this is the only implementation using it right now I am okay with leaving it as-is for now.